controllers/ssp_controller.go

```diff
@@ -680,19 +676,43 @@ func watchClusterResources(builder *ctrl.Builder, sspOperands []operands.Operand
 	)
 }
 
-func watchResources(builder *ctrl.Builder, handler handler.EventHandler, sspOperands []operands.Operand, watchTypesFunc func(operands.Operand) []client.Object, hookFunc handler_hook.HookFunc) {
-	watchedTypes := make(map[reflect.Type]struct{})
+func watchResources(ctrlBuilder *ctrl.Builder, handler handler.EventHandler, sspOperands []operands.Operand, watchTypesFunc func(operands.Operand) []operands.WatchType, hookFunc handler_hook.HookFunc) {
+	// Deduplicate watches
+	watchedTypes := make(map[reflect.Type]operands.WatchType)
 	for _, operand := range sspOperands {
-		for _, t := range watchTypesFunc(operand) {
-			if _, ok := watchedTypes[reflect.TypeOf(t)]; ok {
-				continue
+		for _, watchType := range watchTypesFunc(operand) {
+			key := reflect.TypeOf(watchType.Object)
+			if wt, ok := watchedTypes[key]; ok {
+				watchType.WatchFullObject = watchType.WatchFullObject || wt.WatchFullObject
 			}
+			watchedTypes[key] = watchType
+		}
+	}
 
-			builder.Watches(
-				&source.Kind{Type: t},
-				handler_hook.New(handler, hookFunc),
-			)
-			watchedTypes[reflect.TypeOf(t)] = struct{}{}
+	for _, watchType := range watchedTypes {
+		var predicates []predicate.Predicate
+		if !watchType.WatchFullObject {
+			predicates = []predicate.Predicate{relevantChangesPredicate()}
```

Comment:
Maybe the Predicate could be part of WatchType so each operand could precisely define the predicate for its types? No/nil predicate would indicate `WatchFullObject`.

Reply:
I would rather keep it simple. If sometime we will need different predicates for various operands, we can add the code.